src/quadiron_c.cpp

```diff
@@ -0,0 +1,388 @@
+
+    return 0;
+}
+
+int quadiron_fnt32_reconstruct(
+    struct QuadironFnt32* fecp,
+    uint8_t** data,
+    uint8_t** parity,
+    int* missing_idxs,
+    unsigned int destination_idx,
+    size_t block_size)
+{
+    quadiron::fec::RsFnt<uint32_t>* fec =
+        reinterpret_cast<quadiron::fec::RsFnt<uint32_t>*>(fecp);
+    std::vector<uint8_t*> data_vec(fec->n_data);
+    std::vector<uint8_t*> parities_vec(fec->n_outputs);
+    std::vector<quadiron::Properties> parities_props(fec->n_outputs);
+    std::vector<int> missing_idxs_vec(
+        missing_idxs, missing_idxs + fec->code_len);
+    std::vector<bool> wanted_data_vec(fec->n_data, false);
+    std::vector<bool> wanted_idxs_vec(fec->n_outputs, false);
+    int metadata_size = quadiron_fnt32_get_metadata_size(fecp, block_size);
+    bool res;
+
+    if (fec->type == quadiron::fec::FecType::SYSTEMATIC) {
+        for (unsigned i = 0; i < fec->n_data; i++) {
+            data_vec[i] = data[i] + metadata_size;
+        }
+        for (unsigned i = 0; i < fec->n_parities; i++) {
+            if (!missing_idxs[fec->n_data + i]) {
+                uint32_t* metadata = reinterpret_cast<uint32_t*>(parity[i]);
+                int ret = parities_props[i].fnt_deserialize(
+                    metadata, metadata_size / 4);
+                if (ret == -1)
+                    return -1;
+            }
+            parities_vec[i] = parity[i] + metadata_size;
+        }
+    } else {
+        for (unsigned i = 0; i < fec->n_data; i++) {
+            if (!missing_idxs[i]) {
+                uint32_t* metadata = reinterpret_cast<uint32_t*>(data[i]);
+                int ret = parities_props[i].fnt_deserialize(
+                    metadata, metadata_size / 4);
+                if (ret == -1)
+                    return -1;
+            }
+            parities_vec[i] = data[i] + metadata_size;
+            data_vec[i] = data[i] + metadata_size;
+        }
+        for (unsigned i = 0; i < fec->n_parities; i++) {
+            if (!missing_idxs[fec->n_data + i]) {
+                uint32_t* metadata = reinterpret_cast<uint32_t*>(parity[i]);
+                int ret = parities_props[fec->n_data + i].fnt_deserialize(
+                    metadata, metadata_size / 4);
+                if (ret == -1)
+                    return -1;
+            }
+            parities_vec[fec->n_data + i] = parity[i] + metadata_size;
+        }
+    }
+
+    int need_decode = 0;
+
+    if (fec->type == quadiron::fec::FecType::SYSTEMATIC) {
+        /*
+         * Easy case where the target is a data then simply decode
+         */
+        if (destination_idx < fec->n_data) {
+            wanted_idxs_vec[destination_idx] = true;
+            res = fec->decode_blocks_vertical(
+                data_vec,
+                parities_vec,
+                parities_props,
+                missing_idxs_vec,
+                wanted_idxs_vec,
+                block_size);
+            if (!res) {
+                return -1;
```

Comment:
this code leaks if there is an error because we don't delete allocated buffer @slaperche-scality